### ISSUE TITLE
Current version downloading 0 byte files

### ISSUE THREAD

**Issue by Falord** [Bionus/imgbrd-grabber#630]
### What steps will reproduce the problem?

Well, all I did was try to open a favorites or two, once this version of grabber was opened or even after waiting a couple minutes and trying to get atleast one to load, repeatedly if necessary. 
### What is the expected behavior? What do you get instead?

I did this up to version 4.7.0 since that was the version I last had working on my system and it caused no problems and would all I clicked near immediately. Once I clicked a tag in this version, in this case my favorites, the program would freeze for a few seconds and open it in a tab. Once opened, attempting to either reload that tab or open new tags well freeze the program again or just crash it.
### What version of the program are you using? On what operating system?

Version 4.8.2 this happening on. I'm on Windows 8.  
### Please provide any additional information below

This is more off topic, but I've notice that the more recent installers no longer give the option to pick your location to install grabber and uses the last set location from my old installs. Can the option be re-implemented?

**Comment by MrCrackR** [comment]
Just look at the RAM usage. At about 1.900MB it will crash. That happens if there are many image search results in the open tabs (in total) and if the download query is very long. It reminds me of the 1.900MB margin that crashed firefox in the past, when it was only available in its 32-bit version. But in this case 64-bit wouldn't help much. Somewhere there is a lot of memory wasted. Unfortunately I'm not a C++ dev so I can't help much.

If you want to pick a new location, just uninstall the old installation before installing the new version. If you didn't do the portable method, than the old settings should remain in _C:\Users\UserName\AppData\Local\Bionus\Grabber_ after uninstalling (at least it did so for me).

**Comment by Bionus** [comment]
About crashes, as said on issue #618 this version introduced a few big leaks that I'm currently working on. I'll upload a new version as soon as I get satisfying results. About the freezes, it's also related to the leak issue and I have already a fix on develop, so it should also be fixed as soon as the new version is released.

Sorry about this, this new version introduces some big internal changes, which caused these leaks to appear. I'm currently working on it as fast as I can to get it fixed, so it shouldn't take more than a few days.

> This is more off topic, but I've notice that the more recent installers no longer give the option to pick your location to install grabber and uses the last set location from my old installs. Can the option be re-implemented?

I didn't change anything in the installer, so I guess it just doesn't prompt for the install folder if you have it already installed on your system. So as MrCrackR said, uninstall first. If you want to have multiple versions installed at the same time, you'd be better off using the portable mode.

> Just look at the RAM usage. At about 1.900MB it will crash. That happens if there are many image search results in the open tabs (in total) and if the download query is very long. It reminds me of the 1.900MB margin that crashed firefox in the past, when it was only available in its 32-bit version. But in this case 64-bit wouldn't help much. Somewhere there is a lot of memory wasted. Unfortunately I'm not a C++ dev so I can't help much.

That's also my conclusion for all these crashes, that's why I'm currently re-doing most of the program's memory management. There's currently a 5mo leak per page since version 4.8.0 (my bad), and there has been a 200ko leak per image since a long time ago (causing the crashes on big batch downloads).

**Comment by Bionus** [comment]
After a few fixes, I managed to get rid of the 5mo/page leak, and reduce the 200ko/image one to around 60ko/image.

**Comment by Bionus** [comment]
With the new version, it's still not perfect but already a lot better, as you can see below (stats were done with the Windows task manager on windows 10, using version 4.8.3).

Sorry for the inconvenience! I just uploaded version 4.8.3. Please tell me if it fixed your issues, so that I can know if I still have to investigate or not.

## Download of 2,000 images

* Before download: 21.5 mo
* Beginning of the download: 86.1 mo
* End of the download: 112 mo
* After download: 69.8 mo

10 pages and 2,000 images took 65 mo to store in memory (32 ko per image).
Each image downloaded "leaked" 13 ko.
In the end, 45 mo were leaked.

## Download of 20,000 images

* Before download: 17.4 mo
* Beginning of the download: 458 mo
* After 1,000 images: 487 mo
* After 3,000 images: 506 mo
* After 5,000 images: 508 mo (down from some peaks at 550 mo)
* After 6,000 images: 470 mo
* After 8,000 images: 471 mo
* After 10,000 images: 383 mo
* After 15,000 images: 378 mo
* After 18,000 images: 311 mo
* End of the download: 359 mo
* After download: 239 mo

100 pages and 20,000 images took 440 mo to store in memory (22 ko per image).
In the end, 220 mo were leaked.

What can be easily noticed is that except for the initial storage in RAM of the images' data, the download itself did not need much more even though we had 10 times more images. The further in the download we were, the less it ate RAM (even getting some back each time it recycled the already downloaded images).

In conclusion, there shouldn't be issues as long as you don't go above 100,000 images for now (after which the storage itself of the data will eat all your RAM).

**Comment by Falord** [comment]
You didn't change the installer? That's strange. But from version 4.5.3 back, still ask you to pick an install location regardless of if it was already installed. Most be the older versions were bugged for still asking then!

**Comment by Falord** [comment]
Just checked version 4.8.3. It no longer freezes and moves like normal with crashes, but no longer downloads. It just ignores all in batch downloads. A matter of fact, nothing saves in the new version. Even using 'save' or 'save and close' for me on the image itself produces a file with 0 bytes.

**Comment by MrCrackR** [comment]
For me it now works perfectly except for images from the source rule34.paheal.net, where it doesn't save the images but some HTML-files.
![just an example for such a file](https://cloud.githubusercontent.com/assets/1807632/19332571/b3357554-90ed-11e6-9b6b-1b87f8700725.jpg)

**Comment by Bionus** [comment]
> For me it now works perfectly except for images from the source rule34.paheal.net, where it doesn't save the images but some HTML-files.

Yeah, they broke their RSS API and for some reason it's above the Regex API in the new version. A workaround is to change the site's source order to put the Regex on top (requires a restart).

> version 4.8.3 [...] no longer downloads. It just ignores all in batch downloads. A matter of fact, nothing saves in the new version. Even using 'save' or 'save and close' for me on the image itself produces a file with 0 bytes.

> I can confirm now as well, that there are some files, that end up as 0-byte-files on my HDD. Not all of them, but a few.

I'll investigate, but I think some logs may help for this, until I can reproduce the issue. Is it totally random? (all of Falord's image seem to break, while only a few of MrCrackR's have this issue)

**Comment by Falord** [comment]
The batch downloads are work now. After restarting my computer they started working. But all images that are saved using "save" and "save and close" all produce 0 byte files for me still on version 4.8.3.  Hope these are the right logs.

[main.zip](https://github.com/Bionus/imgbrd-grabber/files/529679/main.zip)

**Comment by MrCrackR** [comment]
The only error I get is:
> [17:33:33.254] Error parsing RSS file: letter is expected (4 - 53).
> [17:33:33.254] Loading using Rss failed. Retry using Html.

This log is written, when the 0-byte file is created:
> [17:41:13.950] Loading image from http://simg4.*booru.com//images/86/9d/*.jpg
> [17:41:14.156] Image received from http://simg4.*booru.com//images/86/9d/*.jpg
> [17:41:14.156] Saving image in D:\Files\\\*\\\*booru.com\3192782.jpg

They are only created when directly viewing them inside Grabber (on every source), not when downloading them with the function in the Downloads-Tab.

**Comment by barryanders** [comment]
I'm also having this issue while testing with safebooru.

![screen shot 2016-10-19 at 12 34 08 pm](https://cloud.githubusercontent.com/assets/5648875/19530222/93483226-95f8-11e6-8b59-cdadb360536c.png)

**Comment by Bionus** [comment]
The issue should be fixed in the `develop` branch, and the fix will be included in the next release. :smile:
